### PR TITLE
Ka/python function

### PR DESCRIPTION
**Pull request by KamalAbboud**
Implementation of the multiprocessing mechanism for func_sample argument
in PythonFunction class.

### CODE REVIEW

**Comment by jschueller** [comment]
Sorry but this is still not clear to me what this PR does, I understand it changes the way the exec_sample argument is used, but I dont see what it changes exactly except that it uses numpy, as there was already a multiprocess implementation using map_async. Maybe you need to provide benchmarks as well.

File: python/test/t_PythonFunction_multiprocessing.expout [hunk 1]
```diff
@@ -0,0 +1,2 @@
+T sequential =  62 (s)
```
Comment (jschueller) [inline python/test/t_PythonFunction_multiprocessing.expout]: dont expect the time to be the same across different build hosts!

Comment (KamalAbboud) [reply python/test/t_PythonFunction_multiprocessing.expout]: Hello Julien, You are right there was a regression due to forgotten a case. I pushed a correction and i am preparing a benchmark with more explanation for the next pull request. Thanks

**Comment by KamalAbboud** [comment]
Different cases in PythonFunction to distinguish
================================================

+---+----------+-------------+--------+--------------+-----------------------------+
| # |   func   | func_sample | n_cpus | _exec        | _exec_sample                |
+===+==========+=============+========+==============+=============================+
| 1 |   func   |    None     |  None  | func         | instance.p2s                |
+---+----------+-------------+--------+--------------+-----------------------------+
| 2 |   None   | func_sample |  None  | instance.s2p | func_sample                 |
+---+----------+-------------+--------+--------------+-----------------------------+
| 3 |   func   | func_sample |  None  | func         | func_sample                 |
+---+----------+-------------+--------+--------------+-----------------------------+
| 4 |   func   |    None     | n_cpus | func         | _exec_sample_mp_func_       |
+---+----------+-------------+--------+--------------+-----------------------------+
| 5 |   None   | func_sample | n_cpus | instance.s2p | _exec_sample_mp_func_sample |
+---+----------+-------------+--------+--------------+-----------------------------+
| 6 |   func   | func_sample | n_cpus | func         | _exec_sample_mp_func_sample |
+---+----------+-------------+--------+--------------+-----------------------------+
instance.s2p = instance._exec_point_on_exec_sample
instance.p2s = instance._exec_sample
_exec_sample_mp_func = _exec_sample_multiprocessing_func(func, n_cpus)
_exec_sample_mp_func_sample =
_exec_sample_multiprocessing_func_sample(func_sample, n_cpus)

Cases inpacted by this project
==============================

- case 4: we just renamed _exec_sample_multiprocessing -> _exec_sample_multiprocessing_func.
- case 5/6: were added to handel multiprocessing of func_sample.

For information, in the master the case 5 was handled as case 2 (n_cpus was ignored as = None).
And the case 6 was handled like case 4.


Benchemark Results
==================
n_cpus = 4

func_sample without multiprocessing (cases 2 and 3)
---------------------------------------------------

+---+----------+-------------------+
| # |   master | ka/pythonFunction |
+===+==========+===================+
| 2 |   12.4s  |      12.4s        |
+---+----------+-------------------+
| 3 |   12.4s  |      12.4s        |
+---+----------+-------------------+


func_sample with multiprocessing (cases 5 and 6)
------------------------------------------------

+---+----------+-------------------+
| # |   master | ka/pythonFunction |
+===+==========+===================+
| 5 |   12.4s  |       5.3s        |
+---+----------+-------------------+
| 6 |   12.4s  |       5.3s        |
+---+----------+-------------------+



#! /usr/bin/env python

from __future__ import print_function
import openturns as ot
import numpy as np
from time import time, sleep
import multiprocessing
n_cpus = multiprocessing.cpu_count()
 
ot.TESTPREAMBLE()

nvar = 100
ng   = 1 # Gb

nsim = int(ng*1e9/nvar/8) # double precision
rho  = 0.00001 # rho s/ point evaluation
T    = rho*nsim


def pointFuncPython(x):
    from time import time
    y = [0]
    y[0] = 1 - x[0]
    return  y

def sampleFuncPython(X):
    from time import time
    XArray = np.array(X, copy=False)
    res = 1 - XArray[:,0]
    sleep(int(rho * len(X)))
    return  res.reshape(len(XArray), 1)      

def createFunction2():
    return ot.PythonFunction(nvar, 1, func_sample=sampleFuncPython)

def createFunction3():
    return ot.PythonFunction(nvar, 1, func=pointFuncPython, func_sample=sampleFuncPython)

def createFunction5():
    return ot.PythonFunction(nvar, 1, func_sample=sampleFuncPython, n_cpus=n_cpus)

def createFunction6():
    return ot.PythonFunction(nvar, 1, func=pointFuncPython, func_sample=sampleFuncPython, n_cpus=n_cpus)

print("Benchmark:")

print(" T sequential = ", int(T), "(s)")
print(" Mem          = ", int(ng), "(Gb)")
print(" nvar         = ", nvar)
print(" nsim         = ", nsim)
print("\n")

print("====================================================")
print("without multiprocessing (cases 2, 3)")
print("====================================================")

print("\n")
print("func = None, n_cpus = None")
print("--------------------------")
myfun = createFunction2()
sample = np.linspace(1.0 / nsim, 1.0, num=nsim)
input = np.repeat(np.expand_dims(sample, axis=1), myfun.getInputDimension(), axis=1)
t0 = time()
out = myfun(input)
t1 = time()
print("   t1 - t0                 : ", t1 - t0 )
T2 = t1 - t0

print("\n")
print("func != None, n_cpus = None")
print("---------------------------")
myfun = createFunction3()
sample = np.linspace(1.0 / nsim, 1.0, num=nsim)
input = np.repeat(np.expand_dims(sample, axis=1), myfun.getInputDimension(), axis=1)
t0 = time()
out = myfun(input)
t1 = time()
print("   t1 - t0                 : ", t1 - t0 )
T3 = t1 - t0

print("\n")


print("\n")
print("====================================================")
print("with multiprocessing (cases 5, 6)"    )
print("====================================================")
print("\n")


print("func = None, n_cpus > 1")
print("-----------------------")
myfun = createFunction5()
sample = np.linspace(1.0 / nsim, 1.0, num=nsim)
input = np.repeat(np.expand_dims(sample, axis=1), myfun.getInputDimension(), axis=1)
t0 = time()
out = myfun(input)
t1 = time()
print("   t1 - t0                   : ", t1 - t0 )
print("   Speed factor              : ",T2/(t1 - t0)  )
print("\n")


print("func != None, n_cpus > 1")
print("------------------------")
myfun = createFunction6()
sample = np.linspace(1.0 / nsim, 1.0, num=nsim)
input = np.repeat(np.expand_dims(sample, axis=1), myfun.getInputDimension(), axis=1)
t0 = time()
out = myfun(input)
t1 = time()
print("    t1 - t0                 : ", t1 - t0 )
print("    Speed factor            : ", T3/(t1 - t0)  )

**Comment by regislebrun** [comment]
+1

**Comment by regislebrun** [comment]
@jschueller @sofianehaddad Still something to change here?

**Comment by jschueller** [comment]
I'll look into it monday.

**Pull request merged**